### PR TITLE
[MEAR-194] Use ear archiver instead of jar archiver

### PR DESCRIPTION
**Pull request by santik**
In plugin instead of correct EAR archiver JAR archiver was used.
In this PR application will use correct archiver.

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 1]
```diff
@@ -692,9 +693,9 @@ public String getMappedFileName( String pName )
      * 
      * @return the archiver
      */
-    protected JarArchiver getJarArchiver()
+    protected EarArchiver getEarArchiver()
```
Comment (pzygielo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:696]: Perhaps `link` in doc could be updated and import of `JarAchiver` removed?

Comment (santik) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:695]: done

**Comment by pzygielo** [comment]
On fb2a47cf8cdbc23639b49d6f892a9a0e2fc766d2 I got
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-ear-plugin:3.1.0-SNAPSHOT:ear (default-ear) on project maven-ear-plugin-test-project-054: Error assembling EAR: Deployment descriptor: .../maven-ear-plugin/target/test-classes/projects/project-054/target/maven-ear-plugin-test-project-054-99.0/META-INF/application.xml does not exist. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-ear-plugin:3.1.0-SNAPSHOT:ear (default-ear) on project maven-ear-plugin-test-project-054: Error assembling EAR
...
[ERROR] Errors: 
[ERROR]   EarMojoIT.testProject054:652->AbstractEarPluginIT.doTestProject:184->AbstractEarPluginIT.doTestProject:169->AbstractEarPluginIT.doTestProject:144->AbstractEarPluginIT.executeMojo:128->AbstractEarPluginIT.executeMojo:97 » Verification
```

**Comment by santik** [comment]
> On [fb2a47c](https://github.com/apache/maven-ear-plugin/commit/fb2a47cf8cdbc23639b49d6f892a9a0e2fc766d2) I got
> 
> ```
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-ear-plugin:3.1.0-SNAPSHOT:ear (default-ear) on project maven-ear-plugin-test-project-054: Error assembling EAR: Deployment descriptor: .../maven-ear-plugin/target/test-classes/projects/project-054/target/maven-ear-plugin-test-project-054-99.0/META-INF/application.xml does not exist. -> [Help 1]
> org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-ear-plugin:3.1.0-SNAPSHOT:ear (default-ear) on project maven-ear-plugin-test-project-054: Error assembling EAR
> ...
> [ERROR] Errors: 
> [ERROR]   EarMojoIT.testProject054:652->AbstractEarPluginIT.doTestProject:184->AbstractEarPluginIT.doTestProject:169->AbstractEarPluginIT.doTestProject:144->AbstractEarPluginIT.executeMojo:128->AbstractEarPluginIT.executeMojo:97 » Verification
> ```

You are right. Sometimes we still need to use JAR archiver. Updated the code.

**Comment by hboutemy** [comment]
question: what does an ear archiver do differently from a jar archiver that brings concrete benefit, please?

**Comment by santik** [comment]
@hboutemy it is described in the ticket as showing incorrect message
however I think it is better to use correct implementation instead of generic jar archiver when it is possible

**Comment by hboutemy** [comment]
ok, now I get that the message is written by the archiver, not by the plugin
adn from a pure theoretical perspective, a more dedicated archiver can help in some cases but given the plugin worked perfectly with simple jar, I need to clarify if any ear archiver feature won't conflict
but sure, I'll rework the provided patch to fix the current conflict with master, if you don't beat me at it :)

**Comment by santik** [comment]
@hboutemy sure! I've just resolved conflicts, but if you have anything to add- do it

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 2]
```diff
@@ -289,12 +296,25 @@ public void execute()
...
+            JarArchiver theArchiver;
+            if ( ddFile.exists() )
+            {
+                final EarArchiver earArchiver = getEarArchiver();
+                earArchiver.setAppxml( ddFile );
+                theArchiver = earArchiver;
+            }
+            else
+            {
+                theArchiver = getJarArchiver();
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:313]: given there is a test later that checks and fail if the descriptor does not exist, it seems this case is not really useful, isn't it?

Comment (santik) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:313]: file is not required for the jar archiver, but required for ear

Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:313]: I tested and found that starting with JavaEE 5, descriptor is not required: but currently, EarArchiver is implemented in a way that does not support this condition
ok, now I see how we must either improve EarArchiver, either do the workaround you did: I'll merge and document the workaround

Comment (santik) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: ok
I fixed conflicts again

**Comment by hboutemy** [comment]
merged and updated to document the unexpectedly complex logic for an issue that we supposed was up-for-grabs... :)

thank you everybody for your help and efforts

**Comment by jira-importer** [comment]
Resolve #369

**Pull request closed without merging**